File: examples/csharp/Recipes/CreateBins.cs
// Recipe: Create Bins (Bulk)
// Docs:   docs/recipes/create-bins.md
//
// Bulk-create warehouse bins with the BinLocation service — one transaction
// per bin, tens per POST, verified in production at hundreds of bins per run.
//
// Key rules (all verified live — see the recipe page):
//   - IgnoreDisabled: true is mandatory and must be at the payload top level;
//     inside a Transaction it is silently ignored (frozen_flag is disabled).
//   - Pass codes, not uids (bin_type "SHELF", zone "ZONE-A").
//   - Flags are ON/OFF on the form but stored Y/N in dbo.bin — convert when
//     cloning constants from a "twin" bin.
//   - Don't send master_bin_flag — P21 auto-sets it.
//   - Re-running is safe if you skip (bin_id, location_id) pairs that exist.

using Newtonsoft.Json.Linq;
using P21Examples.Common;
using static P21Examples.Recipes.RecipeHelpers;

namespace P21Examples.Recipes;

public static class CreateBins
{
    private const string CompanyId = "ACME";
    private const string LocationId = "10";
    private const int BatchSize = 20;  // tens of transactions per POST is fine and fast

    public static async Task RunAsync()
    {
        Console.WriteLine("Recipe - Create Bins Bulk (BinLocation)");
        Console.WriteLine(new string('=', 60));

        using var client = await P21Client.CreateAsync();

        var newBinIds = new[] { "A01-02-01", "A01-02-02", "A01-02-03", "A01-02-04" };

        // Constants cloned from a "twin" bin of the same bin_type at this location.
        // Flags come back Y/N from the database — convert to ON/OFF for the form.
        var twin = new (string Name, string Value)[]
        {
            ("bin_type", "SHELF"),
            ("putaway_zone_id", "ZONE-A"), ("pick_zone_id", "ZONE-A"),
            ("bin_length", "10"), ("bin_width", "10"), ("bin_height", "11"),
            ("warehouse_sequence", "1"), ("putaway_zone_sequence", "1"), ("pick_zone_sequence", "1"),
            ("max_unique_items", "0"),
            ("pick_locked_flag", "OFF"), ("put_locked_flag", "OFF"),
            ("full_flag", "OFF"), ("frozen_flag", "OFF"),
            ("consolidation_bin_flag", "OFF"), ("stage_bin_flag", "OFF"), ("door_bin_flag", "OFF"),
        };

        // ------------------------------------------------------------------
        // Read-only: skip-existing check via the p21_view_bin view
        // (the raw bin table isn't always exposed via OData)
        // ------------------------------------------------------------------
        var existingRows = (JArray)(await client.OData.QueryViewAsync(
            "p21_view_bin",
            select: "bin_id",
            filter: $"location_id eq {LocationId}"))["value"]!;
        var existing = existingRows.Select(r => (string)r["bin_id"]!).ToHashSet();

        var toCreate = newBinIds.Where(b => !existing.Contains(b)).ToList();
        Console.WriteLine(
            $"\n{newBinIds.Length - toCreate.Count} already exist, creating {toCreate.Count}");
        if (toCreate.Count == 0)
            return;

        var batches = toCreate.Chunk(BatchSize).ToList();
        foreach (var batch in batches)
        {
            var preview = BuildBatchPayload(batch, twin);
            PrintPayload($"Batch payload ({batch.Length} bin(s))", preview);
        }

        if (!ConfirmExecute())
            return;

        // ------------------------------------------------------------------
        // Write: one POST per batch, per-transaction result checks
        // ------------------------------------------------------------------
        foreach (var batch in batches)
        {
            var payload = BuildBatchPayload(batch, twin);
            var result = await client.Transaction.CreateAsync(payload);
            Console.WriteLine();
            CheckResult(result);

            // Transactions pass/fail independently — check each one
            var txns = result.Raw?["Results"]?["Transactions"] as JArray ?? new JArray();
            foreach (var (binId, txn) in batch.Zip(txns))
                if ((string?)txn["Status"] != "Passed")
                    Console.WriteLine($"  FAILED: {binId}");
        }

        // ------------------------------------------------------------------
        // Verify: read the new bins back through p21_view_bin
        // ------------------------------------------------------------------
        Console.WriteLine("\nVerify via p21_view_bin:");
        Console.WriteLine(new string('-', 50));
        foreach (var binId in toCreate)
        {
            var rows = (JArray)(await client.OData.QueryViewAsync(
                "p21_view_bin",
                filter: $"location_id eq {LocationId} and bin_id eq '{binId}'"))["value"]!;
            Console.WriteLine(rows.Count > 0
                ? $"  {binId}: exists (bin_type={rows[0]["bin_type"]})"
                : $"  {binId}: NOT FOUND");
        }
        Console.WriteLine(
            "\n  Compare field-for-field against the twin after the first run" +
            " (remember the Y/N <-> ON/OFF flag mapping).");
    }

    private static JObject BuildBatchPayload(
        string[] batch, (string Name, string Value)[] twin)
    {
        return new JObject
        {
            ["Name"] = "BinLocation",
            ["UseCodeValues"] = false,
            ["IgnoreDisabled"] = true, // TOP LEVEL — inside a Transaction it is silently ignored
            ["Transactions"] = new JArray(batch.Select(b => BuildBinTransaction(b, twin))),
        };
    }

    /// <summary>One Transaction object per bin (keys first, then the twin's constants).</summary>
    private static JObject BuildBinTransaction(string binId, (string Name, string Value)[] twin)
    {
        var edits = new JArray
        {
            Edit("company_id", CompanyId),
            Edit("location_id", LocationId),
            Edit("bin_id", binId),
        };
        foreach (var (name, value) in twin)
            edits.Add(Edit(name, value));

        return new JObject
        {
            ["Status"] = "New",
            ["DataElements"] = new JArray
            {
                new JObject
                {
                    ["Name"] = "FORM.form", ["Type"] = "Form",
                    ["Keys"] = new JArray("company_id", "location_id", "bin_id"),
                    ["Rows"] = new JArray(new JObject { ["Edits"] = edits }),
                }
            },
        };
    }
}

File: examples/python/recipes/create_bins.py
"""
Create Bins in Bulk (BinLocation)

Bulk-create warehouse bins with the BinLocation service -- one transaction
per bin, tens per POST. Constants are cloned from a "twin" bin of the same
bin_type at the location; bins that already exist are skipped so re-running
is safe. IgnoreDisabled: true at the payload TOP LEVEL is mandatory
(frozen_flag and other system columns are disabled on the bin form).

Mirrors: docs/recipes/create-bins.md

Usage:
    python examples/python/recipes/create_bins.py            # dry run (default)
    python examples/python/recipes/create_bins.py --execute  # skip-existing check, POST + verify
"""

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

import httpx
from common.auth import get_token, get_auth_headers, get_ui_server_url
from common.config import load_config

import warnings
warnings.filterwarnings("ignore")

# --- Configuration (generic placeholders -- substitute your own) ------------
COMPANY_ID = "ACME"
LOCATION_ID = "10"
NEW_BIN_IDS = ["A01-02-01", "A01-02-02", "A01-02-03", "A01-02-04"]
BATCH_SIZE = 20  # tens of transactions per POST is fine and fast

# Constants cloned from a "twin" bin of the same bin_type at this location.
# Flags come back Y/N from the database -- convert to ON/OFF for the form.
TWIN = {
    "bin_type": "SHELF",
    "putaway_zone_id": "ZONE-A", "pick_zone_id": "ZONE-A",
    "bin_length": "10", "bin_width": "10", "bin_height": "11",
    "warehouse_sequence": "1", "putaway_zone_sequence": "1", "pick_zone_sequence": "1",
    "max_unique_items": "0",
    "pick_locked_flag": "OFF", "put_locked_flag": "OFF",
    "full_flag": "OFF", "frozen_flag": "OFF",
    "consolidation_bin_flag": "OFF", "stage_bin_flag": "OFF", "door_bin_flag": "OFF",
}


def build_bin_transaction(bin_id: str, location_id: str, twin: dict) -> dict:
    """Build one Transaction object per bin (keys first, then twin constants).

    Args:
        bin_id: New bin ID to create.
        location_id: Stocking location for the bin.
        twin: Field constants cloned from an existing bin of the same type.

    Returns:
        dict: Transaction object for the BinLocation payload.
    """
    edits = [
        {"Name": "company_id", "Value": COMPANY_ID},
        {"Name": "location_id", "Value": location_id},
        {"Name": "bin_id", "Value": bin_id},
    ] + [{"Name": name, "Value": value} for name, value in twin.items()]
    return {
        "Status": "New",
        "DataElements": [{
            "Name": "FORM.form", "Type": "Form",
            "Keys": ["company_id", "location_id", "bin_id"],
            "Rows": [{"Edits": edits}],
        }],
    }


def build_payload(bin_ids: list[str]) -> dict:
    """Wrap bin transactions in the top-level BinLocation payload.

    Args:
        bin_ids: Bin IDs for this batch.

    Returns:
        dict: Complete Transaction API payload (IgnoreDisabled at top level).
    """
    return {
        "Name": "BinLocation",
        "UseCodeValues": False,
        "IgnoreDisabled": True,  # TOP LEVEL -- inside a Transaction it is silently ignored
        "Transactions": [build_bin_transaction(b, LOCATION_ID, TWIN) for b in bin_ids],
    }


def main() -> None:
    """Entry point: dry run prints the payload; --execute skips existing bins,
    POSTs batches, and reads the created bins back via p21_view_bin."""
    parser = argparse.ArgumentParser(
        description="Bulk-create warehouse bins (docs/recipes/create-bins.md)")
    parser.add_argument("--execute", action="store_true",
                        help="Actually POST the transactions (default: dry run)")
    args = parser.parse_args()

    print("Recipe - Create Bins in Bulk (BinLocation)")
    print("=" * 60)

    if not args.execute:
        print("\nDRY RUN - payload that would be POSTed to {ui_server}/api/v2/transaction")
        print("(at execute time, bins that already exist at the location are skipped):")
        print(json.dumps(build_payload(NEW_BIN_IDS), indent=2))
        print("\nRe-run with --execute to POST it.")
        return

    config = load_config()
    token_data = get_token(config)
    headers = get_auth_headers(token_data["AccessToken"])
    ui_server = get_ui_server_url(config.base_url, token_data["AccessToken"], config.verify_ssl)
    print(f"UI Server: {ui_server}")

    # Skip-existing check via the p21_view_bin view (raw bin table isn't always in OData)
    existing_resp = httpx.get(
        f"{config.base_url}/odataservice/odata/view/p21_view_bin",
        params={"$filter": f"location_id eq {LOCATION_ID}", "$select": "bin_id"},
        headers=headers, verify=config.verify_ssl, follow_redirects=True,
    )
    existing_resp.raise_for_status()
    existing = {row["bin_id"] for row in existing_resp.json()["value"]}

    to_create = [b for b in NEW_BIN_IDS if b not in existing]
    print(f"{len(NEW_BIN_IDS) - len(to_create)} already exist, creating {len(to_create)}")

    for start in range(0, len(to_create), BATCH_SIZE):
        batch = to_create[start:start + BATCH_SIZE]
        resp = httpx.post(f"{ui_server}/api/v2/transaction",
                          headers=headers, json=build_payload(batch),
                          verify=config.verify_ssl, follow_redirects=True, timeout=300)
        resp.raise_for_status()
        result = resp.json()

        summary = result["Summary"]
        print(f"Batch {start // BATCH_SIZE + 1}: "
              f"Succeeded={summary['Succeeded']}, Failed={summary['Failed']}")
        if summary["Failed"] > 0:
            for msg in result.get("Messages") or []:
                print(f"  {msg}")
        # Transactions pass/fail independently -- check each one
        for bin_id, txn in zip(batch, (result.get("Results") or {}).get("Transactions") or []):
            if txn["Status"] != "Passed":
                print(f"  FAILED: {bin_id}")

    # --- Verify: read the new bins back through the p21_view_bin view ---
    print("\nVerify (p21_view_bin read-back -- compare field-for-field against the twin;")
    print("flags are stored Y/N in the database vs ON/OFF on the form):")
    for bin_id in to_create:
        check = httpx.get(
            f"{config.base_url}/odataservice/odata/view/p21_view_bin",
            params={"$filter": f"location_id eq {LOCATION_ID} and bin_id eq '{bin_id}'"},
            headers=headers, verify=config.verify_ssl, follow_redirects=True,
        )
        check.raise_for_status()
        rows = check.json()["value"]
        print(f"  {bin_id}: {'FOUND' if rows else 'MISSING'}")


if __name__ == "__main__":
    main()
